pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -128,345 +167,552 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
-		return fmt.Errorf(errPolicyMergeMutate, secret.Name, err)
+	// fail if the secret does not exist
+	// this should never happen because we check this before calling this function
+	if existingSecret.UID == "" {
+		return fmt.Errorf(errUpdateNotFound, secretName)
 	}
 
-	// GVK is missing in the Secret, see:
-	// https://github.com/kubernetes-sigs/controller-runtime/issues/526
-	// https://github.com/kubernetes-sigs/controller-runtime/issues/1517
-	// https://github.com/kubernetes/kubernetes/issues/80609
-	// we need to manually set it before doing a Patch() as it depends on the GVK
-	gvks, unversioned, err := r.Scheme.ObjectKinds(secret)
-	if err != nil {
-		return err
-	}
-	if !unversioned && len(gvks) == 1 {
-		secret.SetGroupVersionKind(gvks[0])
+	// set the binding reference to the secret
+	// https://github.com/external-secrets/external-secrets/pull/2263
+	es.Status.Binding = v1.LocalObjectReference{Name: secretName}
+
+	// mutate a copy of the existing secret with the mutation function
+	updatedSecret := existingSecret.DeepCopy()
+	if err := mutationFunc(updatedSecret); err != nil {
+		return fmt.Errorf(errMutate, updatedSecret.Name, err)
 	}
 
-	if equality.Semantic.DeepEqual(existing, secret) {
+	// if the secret does not need to be updated, return early
+	if equality.Semantic.DeepEqual(existingSecret, updatedSecret) {
 		return nil
 	}
-	// Cleaning up Managed fields manually as to keep patch coherence
-	secret.ObjectMeta.ManagedFields = nil
-	// we're not able to resolve conflicts so we force ownership
-	// see: https://kubernetes.io/docs/reference/using-api/server-side-apply/#using-server-side-apply-in-a-controller
-	if err := r.Client.Patch(ctx, secret, client.Apply, client.FieldOwner(fqdn), client.ForceOwnership); err != nil {
-		return fmt.Errorf(errPolicyMergePatch, secret.Name, err)
+
+	// if the existing secret is immutable, we can only update the object metadata
+	if existingSecret.Immutable != nil && *existingSecret.Immutable {
+		// check if the metadata was changed
+		metadataChanged := !equality.Semantic.DeepEqual(existingSecret.ObjectMeta, updatedSecret.ObjectMeta)
+
+		// check if the immutable data/type was changed
+		var dataChanged bool
+		if metadataChanged {
+			// update the `existingSecret` object with the metadata from `updatedSecret`
+			// this lets us compare the objects to see if the immutable data/type was changed
+			existingSecret.ObjectMeta = *updatedSecret.ObjectMeta.DeepCopy()
+			dataChanged = !equality.Semantic.DeepEqual(existingSecret, updatedSecret)
+		} else {
+			// we know there was some change in the secret (or we would have returned early)
+			// we know the metadata was NOT changed (metadataChanged == false)
+			// so, the only thing that could have changed is the immutable data/type fields
+			dataChanged = true
+		}
+
+		// because we use labels and annotations to keep track of the secret,
+		// we need to update the metadata, even if the data is immutable
+		if metadataChanged {
```

Comment:
You could do this in the previous if, no need to do a second if call; since this action is independent of the `dataChanged` bool that is created earlier as you write.